integrations/opsgenie/client.py

```diff
@@ -0,0 +1,153 @@
+        try:
+            response = await self.http_client.get(url=url, params=query_params)
+            response.raise_for_status()
+            return response.json()
+
+        except httpx.HTTPStatusError as e:
+            logger.error(
+                f"HTTP error with status code: {e.response.status_code} and response text: {e.response.text}"
+            )
+            raise
+
+    async def get_paginated_resources(
+        self, resource_type: ObjectKind
+    ) -> AsyncGenerator[list[dict[str, Any]], None]:
+        if cache := event.attributes.get(resource_type.value):
+            yield cache
+            return
+        api_version = await self.get_resource_api_version(resource_type)
+        url = f"{self.api_url}/{api_version}/{resource_type.value}s"
+        pagination_params: dict[str, Any] = {"limit": PAGE_SIZE}
+
+        while url:
+            try:
+                response = await self._get_single_resource(
+                    url=url, query_params=pagination_params
+                )
+                yield response["data"]
+
+                url = response.get("paging", {}).get("next")
+            except httpx.HTTPStatusError as e:
+                logger.error(
+                    f"HTTP error with status code: {e.response.status_code} and response text: {e.response.text}"
+                )
+                raise
+
+    async def get_alert(self, identifier: str) -> dict[str, Any]:
+        api_version = await self.get_resource_api_version(ObjectKind.ALERT)
+        url = f"{self.api_url}/{api_version}/alerts/{identifier}"
+        alert_data = (await self._get_single_resource(url))["data"]
+        return await self.get_related_incident_by_alert(alert_data)
+
+    async def get_oncall_team(self, identifier: str) -> dict[str, Any]:
+        cache_key = f"{ObjectKind.TEAM}-{identifier}"
+        if cache := event.attributes.get(cache_key):
+            return cache
+        api_version = await self.get_resource_api_version(ObjectKind.TEAM)
+        url = f"{self.api_url}/{api_version}/teams/{identifier}"
+        oncall_team = (await self._get_single_resource(url))["data"]
+        event.attributes[cache_key] = oncall_team
+        return oncall_team
+
+    async def get_oncall_user(self, schedule_identifier: str) -> dict[str, Any]:
+        api_version = await self.get_resource_api_version(ObjectKind.SCHEDULE)
+        url = f"{self.api_url}/{api_version}/schedules/{schedule_identifier}/on-calls?flat=true"
+        return (await self._get_single_resource(url))["data"]
+
+    async def get_cached_resources(
+        self, resource_type: ObjectKind
+    ) -> list[dict[str, Any]]:
+        resources_list = []
+        if cache := event.attributes.get(resource_type.value):
+            return cache
+        async for batch_data in self.get_paginated_resources(resource_type):
+            resources_list.extend(batch_data)
+        event.attributes[resource_type.value] = resources_list
+        return resources_list
```

Comment:
I think this function is over complicating things
just get and store the cache in the `get_paginated_resources` function